File: Submariner/SBRepeatIconTransformer.m
//
//  SBRepeatIconTransformer.m
//  Submariner
//
//  Created by Calvin Buckley on 2022-05-15.
//  Copyright © 2022 Calvin Buckley. All rights reserved.
//

#import "SBRepeatIconTransformer.h"

@implementation SBRepeatIconTransformer


+ (Class)transformedValueClass {
    return [NSImage class];
}

+ (BOOL)allowsReverseTransformation {
    return NO;
}

- (id)transformedValue:(id)value {
    SBPlayerRepeatMode repeatMode = (SBPlayerRepeatMode)[value integerValue];
    NSString *name, *desc;
    if (repeatMode == SBPlayerRepeatNo) {
        // XXX: Make it tinted instead
        name = @"repeat.circle";
        desc = @"Not Repeating";
    } else if (repeatMode == SBPlayerRepeatOne) {
        name = @"repeat.1.circle.fill";
        desc = @"Repeat One";
    } else if (repeatMode == SBPlayerRepeatAll) {
        name = @"repeat.circle.fill";
        desc = @"Repeat All";
    }
    return [NSImage imageWithSystemSymbolName: name accessibilityDescription: desc];
}

@end

File: Submariner/SBRepeatIconTransformer.swift
//
//  SBRepeatIconTransformer.swift
//  Submariner
//
//  Created by Calvin Buckley on 2023-04-10.
//  Copyright © 2023 Submariner Developers. All rights reserved.
//

import Cocoa

@objc(SBRepeatIconTransformer) class SBRepeatIconTransformer: ValueTransformer {
    override class func allowsReverseTransformation() -> Bool {
        false
    }
    
    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }

    override func transformedValue(_ value: Any?) -> Any? {
        // value is NSNumber-based enum
        if let modeRaw = value as? Int, let mode = SBPlayerRepeatMode(rawValue: modeRaw) {
            switch (mode) {
            case .one:
                return NSImage.init(systemSymbolName: "repeat.1", accessibilityDescription: "Repeat One")
            case .all:
                return NSImage.init(systemSymbolName: "repeat", accessibilityDescription: "Repeat All")
            @unknown default:
                return NSImage.init(systemSymbolName: "repeat", accessibilityDescription: "Repeat None")
            }
        }
        return nil
    }
}
